transport-classes-io_uring/src/main/java/io/netty/channel/uring/AbstractIoUringChannel.java

```diff
@@ -250,16 +248,33 @@ private void freeMsgHdrArray() {
     protected void doClose() throws Exception {
         active = false;
 
-        if (registration != null) {
+        if (registration != null && registration.isValid()) {
             if (socket.markClosed()) {
                 int fd = fd().intValue();
                 IoUringIoOps ops = IoUringIoOps.newClose(fd, (byte) 0, nextOpsId());
                 registration.submit(ops);
+
+                boolean cancelConnect = false;
+                try {
+                    ChannelPromise connectPromise = AbstractIoUringChannel.this.connectPromise;
+                    if (connectPromise != null) {
+                        // Use tryFailure() instead of setFailure() to avoid the race against cancel().
+                        connectPromise.tryFailure(new ClosedChannelException());
+                        AbstractIoUringChannel.this.connectPromise = null;
+                        cancelConnect = true;
+                    }
+
+                    cancelConnectTimeoutFuture();
+                } finally {
+                    // It's important we cancel all outstanding connect, write and read operations now so
+                    // we will be able to process a delayed close if needed.
+                    ioUringUnsafe().cancelOps(cancelConnect);
```

Comment:
I think we need to do this before we submit the close, since the fd can be reused as soon as close is completed, and then these operations can end up hitting different operations.

Reply:
good catch!